{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "868a1e03",
   "metadata": {},
   "source": [
    "# 1. METHOD 1: Via Uniprot REST API and `requests`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "50888868",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-30T23:34:54.900142Z",
     "start_time": "2025-11-30T23:34:52.925268Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">sp|P01308|INS_HUMAN Insulin OS=Homo sapiens OX=9606 GN=INS PE=1 SV=1\n",
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "LQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      "✓ Downloaded P01308 to P01308.fasta\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "#! Get the Uniprot request URL\n",
    "api_endpoint = \"https://www.uniprot.org/uniprot/\"\n",
    "protein = \"P01308\" # put the uniprot protein ID here\n",
    "req_url = api_endpoint+protein+\".fasta\"\n",
    "\n",
    "#! Request the protein from Uniprot & print its FASTA sequence\n",
    "response = requests.get(req_url)\n",
    "if response.status_code == 200:\n",
    "    print(response.text)\n",
    "else:\n",
    "    print(\"Something wrong!\")\n",
    "    print(response.status_code)\n",
    "\n",
    "#! Save the FASTA sequence to a file\n",
    "filename = protein+\".fasta\"\n",
    "if response.status_code == 200:\n",
    "    with open(filename, 'w') as file:\n",
    "        file.write(response.text)\n",
    "    print(f\"✓ Downloaded {protein} to {filename}\")\n",
    "else:\n",
    "    print(\"Something wrong!\")\n",
    "    print(response.status_code)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a64dd89",
   "metadata": {},
   "source": [
    "# 2. METHOD 2: via `Entrez` and `SeqIO` module in `Biopython`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "95d6af21",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-30T23:35:48.495171Z",
     "start_time": "2025-11-30T23:35:16.479385Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein ID: sp|P01308.1|INS_HUMAN\n",
      "Description: sp|P01308.1|INS_HUMAN RecName: Full=Insulin; Contains: RecName: Full=Insulin B chain; Contains: RecName: Full=Insulin A chain; Flags: Precursor\n",
      "Length: 110 amino acids\n",
      "Sequence:\n",
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPL...\n",
      "Successfully saved protein ID P01308 to P01308.fasta\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "def retrieve_prot_sequence(protein_id):\n",
    "    \n",
    "    #! Set my email (mandatory for Entrez)\n",
    "    Entrez.email = 'nguyuling@graduate.utm.my'\n",
    "\n",
    "    #! Fetch protein\n",
    "    handle = Entrez.efetch(\n",
    "        db='protein',\n",
    "        id=protein_id,\n",
    "        rettype='fasta',\n",
    "        retmode='text'\n",
    "    )\n",
    "    \n",
    "    #! Read the FASTA record from the handle\n",
    "    record = SeqIO.read(handle, 'fasta')\n",
    "    handle.close()\n",
    "    \n",
    "    #! Extract the sequence string\n",
    "    protein_sequence = str(record.seq)\n",
    "    \n",
    "    #! Print the formatted data\n",
    "    print(\"Protein ID:\", record.id)\n",
    "    print(\"Description:\", record.description)\n",
    "    print(\"Length:\", len(record.seq), \"amino acids\")\n",
    "    print(\"Sequence:\")\n",
    "    print(protein_sequence[:80] + \"...\")\n",
    "\n",
    "    #! Save to file\n",
    "    filename = protein_id + \".fasta\"\n",
    "    with open(filename, \"w\") as file:\n",
    "        SeqIO.write(record, file, \"fasta\")\n",
    "    print(f\"Successfully saved protein ID {protein_id} to {filename}\")\n",
    "    \n",
    "    return protein_sequence\n",
    "\n",
    "# Function Call\n",
    "protein_sequence = retrieve_prot_sequence(\"P01308\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "83b498e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein ID: sp|P01308.1|INS_HUMAN\n",
      "Description: sp|P01308.1|INS_HUMAN RecName: Full=Insulin; Contains: RecName: Full=Insulin B chain; Contains: RecName: Full=Insulin A chain; Flags: Precursor\n",
      "Length: 110 amino acids\n",
      "Sequence:\n",
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPL...\n",
      "Successfully saved protein ID P01308 to P01308.fasta\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "def retrieve_prot_sequence(protein_id):\n",
    "    \n",
    "    #! Set my email (mandatory for Entrez)\n",
    "    Entrez.email = 'nguyuling@graduate.utm.my'\n",
    "\n",
    "    #! Fetch protein\n",
    "    handle = Entrez.efetch(\n",
    "        db='protein',\n",
    "        id=protein_id,\n",
    "        rettype='fasta',\n",
    "        retmode='text'\n",
    "    )\n",
    "    \n",
    "    #! Read the FASTA record from the handle\n",
    "    record = SeqIO.read(handle, 'fasta')\n",
    "    handle.close()\n",
    "    \n",
    "    #! Extract the sequence string\n",
    "    protein_sequence = str(record.seq)\n",
    "    \n",
    "    #! Print the formatted data\n",
    "    print(\"Protein ID:\", record.id)\n",
    "    print(\"Description:\", record.description)\n",
    "    print(\"Length:\", len(record.seq), \"amino acids\")\n",
    "    print(\"Sequence:\")\n",
    "    print(protein_sequence[:80] + \"...\")\n",
    "\n",
    "    #! Save to file\n",
    "    filename = protein_id + \".fasta\"\n",
    "    with open(filename, \"w\") as file:\n",
    "        SeqIO.write(record, file, \"fasta\")\n",
    "    print(f\"Successfully saved protein ID {protein_id} to {filename}\")\n",
    "    \n",
    "    return protein_sequence\n",
    "\n",
    "# Function Call\n",
    "protein_sequence = retrieve_prot_sequence(\"P01308\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8a917b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c0f1cef4-b603-420d-858f-c054580c77cb",
   "metadata": {},
   "source": [
    "# 3. BASIC SEQUENCE ANALYSIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c211c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Amino Acid Composition Results ---\n",
      "   Amino Acid  Count  Percentage\n",
      "2           L     20   18.181818\n",
      "6           G     12   10.909091\n",
      "1           A     10    9.090909\n",
      "15          E      8    7.272727\n",
      "11          Q      7    6.363636\n",
      "5           P      6    5.454545\n",
      "9           V      6    5.454545\n",
      "13          C      6    5.454545\n",
      "4           R      5    4.545455\n",
      "14          S      5    4.545455\n",
      "16          Y      4    3.636364\n",
      "10          N      3    2.727273\n",
      "17          T      3    2.727273\n",
      "8           F      3    2.727273\n",
      "18          K      2    1.818182\n",
      "0           M      2    1.818182\n",
      "12          H      2    1.818182\n",
      "7           D      2    1.818182\n",
      "3           W      2    1.818182\n",
      "19          I      2    1.818182\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "import pandas as pd\n",
    "\n",
    "def generate_basic_analysis(protein_seq):\n",
    "    \n",
    "    #! Count every unique amino acid character\n",
    "    aa_count = Counter(protein_seq)\n",
    "    #! Count the length of the protein sequence\n",
    "    total = len(protein_seq)\n",
    "    #! Calculate the the percentage of each amino acid\n",
    "    aa_percent = {aa: (count / total) * 100 for aa, count in aa_count.items()}\n",
    "\n",
    "    #! Create the data frame\n",
    "    df = pd.DataFrame({\n",
    "        'Amino Acid' : list(aa_count.keys()),\n",
    "        'Count': list(aa_count.values()),\n",
    "        'Percentage': [aa_percent[aa] for aa in aa_count.keys()]\n",
    "    })\n",
    "\n",
    "    #! Sorting Requirement: Sort the DataFrame by the 'Count' column in descending order\n",
    "    df_sorted = df.sort_values(by='Count', ascending=False)\n",
    "    \n",
    "    return df_sorted\n",
    "\n",
    "#! Function Call\n",
    "df_protein = generate_basic_analysis(protein_sequence) # protein_sequence from previous section\n",
    "print(df_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "582e3feb",
   "metadata": {},
   "source": [
    "# PART 1: Working with Protein Sequences (Beginner)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0404b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUn this cell first to install reuired packages\n",
    "# You only need to run this once\n",
    "\n",
    "!pip install biopython requests pandas matplotlib networkx numpy\n",
    "\n",
    "print(\"\\n✓ All packages installed successfully!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35caa392-11c3-43e6-827f-df3bb37a6818",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e28a25cc-3fe5-446d-bc2c-6591bafd77b8",
   "metadata": {},
   "source": [
    "### 1.1 Understanding Protein Sequences\n",
    "#### Background:\n",
    "Proteins are biological molecules made up of amino acids. Each protein has a unique sequence of amino acids that determines its structure and function. In bioinformatics, we represent proteins as strings of one-letter amino acid codes.\n",
    "\n",
    "#### **The 20 Standard Amino Acids:**\n",
    "![20 standard aa](https://i0.wp.com/www.compoundchem.com/wp-content/uploads/2014/09/20-Common-Amino-Acids.png?fit=2480%2C1754&ssl=1)\n",
    "- A (Ala), C (Cys), D (Asp), E (Glu), F (Phe)\n",
    "- G (Gly), H (His), I (Ile), K (Lys), L (Leu)\n",
    "- M (Met), N (Asn), P (Pro), Q (Gln), R (Arg)\n",
    "- S (Ser), T (Thr), V (Val), W (Trp), Y (Tyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "645d14d3-bda3-4d2b-bc64-31f769ce6cf9",
   "metadata": {},
   "source": [
    "### Use Case 1: Analyzing Insulin Protein\n",
    "**Biological Context:** Insulin is a hormone that regulates blood sugar. It's a small protein, making it perfect for learning.\n",
    "\n",
    "**Data Source:** UniProt Database (https://www.uniprot.org/)\n",
    "- Protein: Human Insulin (UniProt ID: P01308)\n",
    "- Format: FASTA (text format for biological sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6435f459-03d4-4d7d-b992-c580eff2154b",
   "metadata": {},
   "source": [
    "### 1.1 Retrieving Protein Sequence Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff52051-8853-4e10-88b6-457319d58807",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Method 1: via Uniprot REST API and `requests`\n",
    "\n",
    "![](http://demoapp.chat/images/api.png)\n",
    "\n",
    "In this method, we will use the API endpoint of Uniprot, send the request via `requests` library in Python. It's the same with how we surf internet, visiting websites. To visit a website, we need to know the URL. Same in here, to retrieve the data, we need to know the \"link\". \n",
    "\n",
    "In this case there will be a standard general part of the link, a.k.a the **\"API endpoint\"**. For Uniprot, we can use the following API endpoint: `https://www.uniprot.org/uniprot/{protein-id}.fasta` where protein-id refers to any specific ID of protein in Uniprot. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "865ea3ca-503c-4ae5-80c2-0550c195d9ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "api_endpoint = \"https://www.uniprot.org/uniprot/\"\n",
    "protein = \"P01308\" # put the uniprot protein ID here\n",
    "req_url = api_endpoint+protein+\".fasta\"\n",
    "\n",
    "print(req_url)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceca1c60-f96c-484d-a86f-dfec06653bbe",
   "metadata": {},
   "source": [
    "Ok, now we have the request URL ready, we need to send the request to the Uniprot server from Python and retrieve the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "705f7007-d3c7-4d60-a54f-1460ab16b93b",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get(req_url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    print(response.text)\n",
    "else:\n",
    "    print(\"Something wrong!\")\n",
    "    print(response.status_code)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9ca6cbf-9dac-4cca-a2d4-53865dc3410c",
   "metadata": {},
   "source": [
    "If we further enhance the code above, we can save the response text into a fasta file. Then we would have a function to download protein sequence data (in FASTA) in Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8e26ca2-f39e-444d-a304-9e732bc39212",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get(req_url)\n",
    "\n",
    "filename = protein+\".fasta\"\n",
    "\n",
    "if response.status_code == 200:\n",
    "    with open(filename, 'w') as file:\n",
    "        file.write(response.text)\n",
    "    print(f\"✓ Downloaded {protein} to {filename}\")\n",
    "else:\n",
    "    print(\"Something wrong!\")\n",
    "    print(response.status_code)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "816ac537-def7-4aa1-897e-9d9d45206688",
   "metadata": {},
   "source": [
    "#### Method 2: via `Entrez` and `SeqIO` module in `Biopython`\n",
    "\n",
    "Different with the Method 1 above, using `Entrez` and `SeqIO` in `Biopython` technically retrieve the same thing given the same protein ID, but the information retrieved from `Entrez` library can be more flexible. For example, one could extract only the sequence itself by using the `.sequence` attribute from the returned result object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0a533c1-28a2-4cd5-9c96-4b3b5f6ccdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70aa3d3b-f856-47cf-9f0d-7e12e655739c",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = 'cwenghowe@utm.my'\n",
    "protein = \"P01308\"\n",
    "handle = Entrez.efetch(db='protein', id=protein , rettype='fasta', retmode='text')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c633f5-1448-4509-a85b-2458552e60be",
   "metadata": {},
   "outputs": [],
   "source": [
    "record = SeqIO.read(handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f120d161-a86e-4230-884d-05cb7b8ed623",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(record.format('fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8e0517a550f6718",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f70a0cc7-5ac5-40ab-a220-2780b5a6b575",
   "metadata": {},
   "source": [
    "To save the retrieved protein data from `Entrez`, we can use `SeqIO.write()` as follows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ab0a5c7f47b6b990",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = protein+\".fasta\"\n",
    "with open(filename, \"w\") as file:\n",
    "    SeqIO.write(record, file, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fa20044-661f-4e6b-ace9-d69899af821a",
   "metadata": {},
   "source": [
    "#### Practice 1\n",
    "Now create a function named: `retrieve_prot_sequence()` that accepts one parameters, the protein ID. and does not return anything. The function basically takes the protein ID, retrieve via `Entrez` method and store into a FASTA file at the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b399728f33607862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my answer\n",
    "\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "def retrieve_prot_sequence(protein_id):\n",
    "    \n",
    "    # Set your email (mandatory for Entrez)\n",
    "    Entrez.email = 'nguyuling@graduate.utm.my'\n",
    "\n",
    "    # Fetch\n",
    "    handle = Entrez.efetch(\n",
    "        db='protein',\n",
    "        id=protein_id,\n",
    "        rettype='fasta',\n",
    "        retmode='text'\n",
    "    )\n",
    "    \n",
    "    # Read the FASTA record from the handle\n",
    "    record = SeqIO.read(handle, 'fasta')\n",
    "    handle.close()\n",
    "    \n",
    "    # Write the SeqRecord object to the FASTA file\n",
    "    filename = f\"{protein_id}.fasta\" \n",
    "    with open(filename, \"w\") as file:\n",
    "        SeqIO.write(record, file, \"fasta\")\n",
    "        \n",
    "    print(f\"Successfully retrieved and saved protein ID {protein_id} to {filename}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b9ee618-90bd-45fb-ae30-a0b65e4c87c3",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "### 1.2 Parsing Protein Sequence Data\n",
    "For the protein data retrieved from `Entrez` method above, we can further parse the data using `SeqIO.parse()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f005691ffbc40674",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lets read the file P01308.fasta\n",
    "\n",
    "insulin_data = []\n",
    "\n",
    "for record in SeqIO.parse('P01308.fasta', \"fasta\"):\n",
    "    insulin_data.append({\n",
    "        'id': record.id,\n",
    "        'description': record.description,\n",
    "        'sequence': str(record.seq),\n",
    "        'length': len(record.seq)\n",
    "    })\n",
    "\n",
    "print(insulin_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90854a8d-9378-40c0-88a4-fe25a4e0f7dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for info in insulin_data:\n",
    "    print(\"Protein ID:\", info['id'])\n",
    "    print(\"Description:\", info['description'])\n",
    "    print(\"Length:\", info['length'], \"amino acids\")\n",
    "    print(\"Sequence:\")\n",
    "    print(info['sequence'][:80] + \"...\")  # Show first 80 characters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "340f6dfb-587f-419a-a5e7-a3df9bab0902",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "### 1.2 Basic Sequence Analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c541807c-e3a8-46ec-8295-a5850168db65",
   "metadata": {},
   "outputs": [],
   "source": [
    "insulin_sequence = insulin_data[0]['sequence']\n",
    "# Sequence Retrieval: Retrieves the amino acid sequence string from the insulin_data list.\n",
    "# insulin_data[0]: Selects the first item in the list (since the FASTA file only contained one record, this is the only dictionary).\n",
    "# ['sequence']: Accesses the value associated with the key 'sequence' inside that dictionary. This final result is the protein sequence string.\n",
    "\n",
    "print(insulin_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "faaf7a9f-6fcc-4e3c-a5dd-d3257e170a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "# Counter is an efficient tool for tallying the frequency of items in a list or characters in a string.\n",
    "\n",
    "# Count each amino acid\n",
    "aa_count = Counter(insulin_sequence)\n",
    "# It takes the insulin_sequence string and automatically counts every unique amino acid character,\n",
    "# storing the result as a dictionary-like object (e.g., {'M': 1, 'A': 5, ...}).\n",
    "\n",
    "total = len(insulin_sequence)\n",
    "# Uses the len() function to get the total number of amino acids in the sequence.\n",
    "\n",
    "print(aa_count)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64fabf86-4b61-4a0f-bad5-ec45d28f9c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate percentages\n",
    "aa_percent = {aa: (count/total)*100 for aa, count in aa_count.items()}\n",
    "# Calculate Percentages (Dictionary Comprehension): This line uses a dictionary comprehension—a concise way to build a new dictionary.\n",
    "# It loops through every item in aa_count (where aa is the amino acid and count is its frequency).\n",
    "# For each amino acid, it calculates the percentage: (count / total) * 100.\n",
    "# The result is a new dictionary, aa_percent, where keys are amino acids and values are their percentage composition.\n",
    "\n",
    "print(aa_percent)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eae8d29f-ac9d-4216-bc29-7b91a99f848b",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
