{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "26e76b5d",
   "metadata": {},
   "source": [
    "# 1. Protein Analysis Using `ProtParam`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aae8a527",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein ID               : sp|P10308.1|FIBER_BPT3\n",
      "Sequence                 : MANVIKTVLTYQLDGSNRDFNIPFEYLARKFVVVTLIGVDRKVLSINADYRFATRTTISLTKAWGPADGYTTIELRRVTSTTDQLVDFTDGSILRAYDLNVAQIQTMHVAEEARDLTADTIGVNNDGHLDARGRRIVNLANAVDDRDAVPFGQLKTMNQNSWQARNEALQFRNEAETFRNQTEVFKNESGTNATNTKQWRDEANGSRDEAEQFKNTAGQYATSAGNSATTATQSEVNAENSATDADNSRNLAEQHADRLELEADKLGIFNGLAGRIDKGDGTNVYWKGGIHANGRLYLTSDGFDCGQYQQFFGGSAGRYSVMEWGIEKAWLMHVERRERTTAIVDNIQLVVNGHIIAQGGDMTGPLKLQNGHALYLESASDKAQYILSKDGNRNNWHIGRGSDNNNDCTFHSYVYGTNLTLKPDYAVVNKRFHVGQAVVATDGNIQGTKWGRKWLDAYLNDTYVKKTMAWTQVWAAASDSYMGGGSQTDTLHRTCDSATYGLRPETTIGTSSELVLTVSTSFQPRRWLKFQIHSNGRVFKNIADRAATPTAIAVEDV\n",
      "Length                   : 557 amino acids\n",
      "Molecular Weight         : 61800.607 Da\n",
      "Isoelectric Point        : 5.9905\n",
      "Instability Index        : 27.0564\n",
      "Aromaticity              : 0.0898\n",
      "Gravy                    : -0.5506\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import ProtParam\n",
    "\n",
    "Entrez.email = 'nguyuling@graduate.utm.my'\n",
    "try:\n",
    "    handle = Entrez.efetch(\n",
    "        db='protein',\n",
    "        id='P10308',\n",
    "        rettype='fasta',\n",
    "        retmode='text'\n",
    "    )\n",
    "    record = SeqIO.read(handle, 'fasta')\n",
    "    handle.close()\n",
    "except Exception as e:\n",
    "    print(f'Error fetching seqeunce: {e}')\n",
    "    \n",
    "\n",
    "def analyse_protein(record):\n",
    "    analysis = ProtParam.ProteinAnalysis(str(record.seq))\n",
    "    properties = {\n",
    "        'Protein ID' : record.id,\n",
    "        'Sequence' : record.seq,\n",
    "        'Length' : f\"{analysis.length} amino acids\",\n",
    "        'Molecular Weight': f\"{round(analysis.molecular_weight(),4)} Da\",\n",
    "        'Isoelectric Point': round(analysis.isoelectric_point(),4),\n",
    "        'Instability Index': round(analysis.instability_index(),4),\n",
    "        'Aromaticity': round(analysis.aromaticity(),4),\n",
    "        'Gravy': round(analysis.gravy(),4)\n",
    "    }\n",
    "    return properties\n",
    "\n",
    "properties = analyse_protein(record)\n",
    "for key, value in properties.items():\n",
    "    print(f'{key:<25}: {value}')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7782581a",
   "metadata": {},
   "source": [
    "# 2. Comparing Protein Family Members"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3aca2a10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving protein sequence for P01308...\n",
      "Retrieving protein sequence for P01322...\n",
      "Retrieving protein sequence for P01325...\n",
      "Retrieving protein sequence for P01329...\n",
      "\n",
      "✓ Data exported successfully.\n",
      "BATCH ANALYSIS COMPLETE: INSULIN FAMILY PROPERTIES\n",
      "            Protein ID Sequence Length Molecular Weight  Isoelectric Point  Instability Index  Aromaticity  GRAVY (Hydrophobicity)\n",
      " sp|P01308.1|INS_HUMAN 110 amino acids    11980.7866 Da             5.2186            40.3328       0.0818                  0.1927\n",
      "  sp|P01322.1|INS1_RAT 110 amino acids    12420.3803 Da             5.6226            47.3674       0.0909                  0.0173\n",
      "sp|P01325.1|INS1_MOUSE 108 amino acids    12160.0224 Da             5.5226            40.4640       0.0833                  0.0435\n",
      " sp|P01329.2|INS_CAVPO 110 amino acids     12186.945 Da             5.1470            59.1583       0.0727                  0.0109\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import ProtParam\n",
    "import pandas as pd\n",
    "import time\n",
    "\n",
    "Entrez.email = 'nguyuling@graduate.utm.my'\n",
    "\n",
    "proteins = ['P01308', 'P01322', 'P01325', 'P01329']\n",
    "records = []\n",
    "processed = 0\n",
    "current   = 0\n",
    "\n",
    "# Fetch all the protein records & Append them\n",
    "while processed < len(proteins):\n",
    "    print(f'Retrieving protein sequence for {proteins[current]}...')\n",
    "    try:\n",
    "        handle = Entrez.efetch(db='protein', id=proteins[current] , rettype='fasta', retmode='text')\n",
    "        record = SeqIO.read(handle, 'fasta')\n",
    "        records.append(record)\n",
    "        processed += 1\n",
    "        current += 1\n",
    "    except:\n",
    "        print(f'Retry after 5 seconds...')\n",
    "        time.sleep(5)\n",
    "        continue\n",
    "\n",
    "# Save the protein records to a file\n",
    "with open('insulin_family.fasta', 'w') as file:\n",
    "    for record in records:\n",
    "        file.write(record.format('fasta'))\n",
    "print(f\"\\n✓ Data exported successfully.\")\n",
    "\n",
    "# Function to analyse a single protein\n",
    "def analyse_protein(record):\n",
    "    analysis = ProtParam.ProteinAnalysis(str(record.seq))\n",
    "    properties = {\n",
    "        'Protein ID': record.id,\n",
    "        'Sequence Length': f\"{analysis.length} amino acids\",\n",
    "        'Molecular Weight': f\"{round(analysis.molecular_weight(), 4)} Da\",\n",
    "        'Isoelectric Point': round(analysis.isoelectric_point(), 4),\n",
    "        'Instability Index': round(analysis.instability_index(), 4),\n",
    "        'Aromaticity': round(analysis.aromaticity(), 4),\n",
    "        'GRAVY (Hydrophobicity)': round(analysis.gravy(), 4),\n",
    "        # You could add other properties like charge, helix/sheet fraction here\n",
    "    }\n",
    "    return properties\n",
    "    \n",
    "# Function to analyse multiple proteins\n",
    "def analyse_multiple_proteins(insulin_fasta):\n",
    "    records = SeqIO.parse(insulin_fasta, 'fasta')\n",
    "    all_proteins_properties = []\n",
    "    for record in records:\n",
    "        protein_properties = analyse_protein(record)\n",
    "        all_proteins_properties.append(protein_properties)\n",
    "    properties_df = pd.DataFrame(all_proteins_properties)\n",
    "    return properties_df\n",
    "\n",
    "prot_family_analysis = analyse_multiple_proteins('insulin_family.fasta')\n",
    "print(\"BATCH ANALYSIS COMPLETE: INSULIN FAMILY PROPERTIES\")\n",
    "print(prot_family_analysis.to_string(index=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b41575a0",
   "metadata": {},
   "source": [
    "# Bioinformatics II: Computational Analysis of Protein Data\n",
    "## Tutorial Notebook - Protein Sequences, Structures, and Interaction Networks\n",
    "\n",
    "**Learning Path:** Beginner → Intermediate → Advanced\n",
    "\n",
    "---\n",
    "\n",
    "### What You'll Learn Today:\n",
    "1. Recap of **Part 1:** \n",
    "2. **Part 2 (Intermediate):** Understanding protein structures - PDB files and structural properties\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8496c0-7261-421c-bfe4-73a021bfa13d",
   "metadata": {},
   "source": [
    "### Recap of Part 1: Working with protein sequences\n",
    "\n",
    "#### Getting the protein properties from sequence\n",
    "\n",
    "Let's just start from the FASTA file downloaded from the Part I of the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62a74f8-6b71-4644-91cc-37566d948c13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data read successfuly!\n",
      "{'id': 'sp|P01308.1|INS_HUMAN', 'description': 'sp|P01308.1|INS_HUMAN RecName: Full=Insulin; Contains: RecName: Full=Insulin B chain; Contains: RecName: Full=Insulin A chain; Flags: Precursor', 'sequence': 'MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN', 'length': 110}\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "seq_data = []\n",
    "\n",
    "for record in SeqIO.parse('P01308.fasta', 'fasta'):\n",
    "    seq_data.append({\n",
    "        'id': record.id,\n",
    "        'description': record.description,\n",
    "        'sequence': str(record.seq),\n",
    "        'length': len(record.seq)\n",
    "    })\n",
    "\n",
    "print(f'Data read successfuly!')\n",
    "print(seq_data[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb4bd4cf-17f3-4d25-923f-ca703008d316",
   "metadata": {},
   "source": [
    "There is one more functions in `BioPython` library that is very handy in getting the protein properties from the sequence, which is `ProtParam` module. \n",
    "1. Let's import the `ProtParam` sub-module from `SeqUtils` module in `BioPython`\n",
    "2. then we will pass the sequence into the `ProteinAnalysis()` of `ProtParam`\n",
    "3. This will produce an `ProtParam` object that we can use to generate different analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ac6225",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import ProtParam\n",
    "\n",
    "analysis = ProtParam.ProteinAnalysis(seq_data[0]['sequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53184a03-6417-4733-a5fd-5b3ef5f29f03",
   "metadata": {},
   "source": [
    "Using the `analysis` variable, which is an `ProtParam` object, we can do following by calling the functions built in for `ProtParam` object.\n",
    "\n",
    "The following codes will get the following properties:\n",
    "- `length`: the length of the sequence\n",
    "- `molecular_weight()`: the total mass of the protein\n",
    "- `isoelectric_point()`: the pH at which the protein has no net charge\n",
    "- `instability_index()`: Instability index of protein, if <40 is stable, >40 is unstable\n",
    "- `aromaticity()`: Relative frequency of aromatic amino acids\n",
    "- `gravy()`: measure hyropathy of protein, negative = hydrophilic, positive = hydrophobic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d0886a84-8b01-40d0-ad78-e9e1b25de4d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein properties\n",
      "------------------\n",
      "Sequence: MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "Sequence length: 110\n",
      "Molecular weight: 11980.79 Da\n",
      "pI: 5.22\n",
      "Instability index: 40.33\n",
      "Aromaticity: 0.08\n",
      "Gravy (hydropathy): 0.19\n"
     ]
    }
   ],
   "source": [
    "print('Protein properties')\n",
    "print('------------------')\n",
    "print(f'Sequence: {analysis.sequence}')\n",
    "print(f'Sequence length: {analysis.length}')\n",
    "print(f'Molecular weight: {analysis.molecular_weight():.2f} Da')\n",
    "print(f'pI: {analysis.isoelectric_point():.2f}')\n",
    "print(f'Instability index: {analysis.instability_index():.2f}')\n",
    "print(f'Aromaticity: {analysis.aromaticity():.2f}')\n",
    "print(f'Gravy (hydropathy): {analysis.gravy():.2f}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa69b128-a8f8-4b82-89c2-9cb12402e8d2",
   "metadata": {},
   "source": [
    "####\n",
    "---\n",
    "#### Now let's make a custom function that produce analysis for given protein sequence:\n",
    "\n",
    "the function `analyse_protein()` below will take the protein sequence, and perform the `ProteinAnalysis()` and eventually retrieve the information including sequence, length, molecular weight, isoelectric point, instability index, aromaticity and gravy. lastly returned as a `dict` (dictionary) format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a275e59c-ee8d-4968-9d20-ad89510a0cbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sequence': 'MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN',\n",
       " 'Length': 110,\n",
       " 'Molecular Weight': 11980.7866,\n",
       " 'Isoelectric Point': 5.2186,\n",
       " 'Instability Index': 40.3328,\n",
       " 'Aromaticity': 0.0818,\n",
       " 'Gravy': 0.1927}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def analyse_protein(sequence):\n",
    "    analysis = ProtParam.ProteinAnalysis(sequence)\n",
    "    properties = {\n",
    "        'Sequence': analysis.sequence,\n",
    "        'Length': analysis.length,\n",
    "        'Molecular Weight': round(analysis.molecular_weight(),4),\n",
    "        'Isoelectric Point': round(analysis.isoelectric_point(),4),\n",
    "        'Instability Index': round(analysis.instability_index(),4),\n",
    "        'Aromaticity': round(analysis.aromaticity(),4),\n",
    "        'Gravy': round(analysis.gravy(),4)\n",
    "    }\n",
    "    return properties\n",
    "\n",
    "\n",
    "# try to call the function\n",
    "analyse_protein(seq_data[0]['sequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48020586-2284-434e-bcab-fd5207666c67",
   "metadata": {},
   "source": [
    "####\n",
    "---\n",
    "\n",
    "### [PRACTICE] Use Case 2: Comparing Protein Family Members\n",
    "**Biological Context:** Let's analyze multiple proteins from the insulin family to compare their properties.\n",
    "\n",
    "**Data:** We'll create a small dataset of insulin-related proteins\n",
    "\n",
    "- P01325 - insulin mouse\n",
    "- P01308 - insulin human\n",
    "- P01322 - insulin rat\n",
    "- P01329 - insulin guinea pigs\n",
    "\n",
    "1. Run the following code to create the small dataset that consists of four proteins above in Fasta format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c31045ad-9ff9-4af5-b07d-5792eed01043",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving protein sequence for P01308...\n",
      "Retrieving protein sequence for P01322...\n",
      "Retrieving protein sequence for P01325...\n",
      "Retrieving protein sequence for P01329...\n",
      "Data exported successfully.\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import time\n",
    "\n",
    "Entrez.email = 'nguyuling@graduate.utm.my'\n",
    "\n",
    "proteins = ['P01308', 'P01322', 'P01325', 'P01329']\n",
    "\n",
    "records = []\n",
    "\n",
    "processed = 0\n",
    "current   = 0\n",
    "\n",
    "while processed < len(proteins):\n",
    "    print(f'Retrieving protein sequence for {proteins[current]}...')\n",
    "    try:\n",
    "        handle = Entrez.efetch(db='protein', id=proteins[current] , rettype='fasta', retmode='text')\n",
    "        record = SeqIO.read(handle, 'fasta')\n",
    "        records.append(record)\n",
    "        processed += 1\n",
    "        current += 1\n",
    "    except:\n",
    "        print(f'Retry after 5 seconds...')\n",
    "        time.sleep(5)\n",
    "        continue\n",
    "\n",
    "with open('insulin_family.fasta', 'w') as file:\n",
    "    for record in records:\n",
    "        file.write(record.format('fasta'))\n",
    "\n",
    "print(\"Data exported successfully.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b019961c-2dc9-417e-af9e-46519a930924",
   "metadata": {},
   "source": [
    "2. Then create a function called `analyse_multiple_proteins()` and pass the fasta file as the input parameter.\n",
    "3. The function should follows the steps below:\n",
    "   1. Read the FASTA file with multiple proteins and parse the proteins sequences into a variable called `sequences`\n",
    "   2. For every protein sequence in the `sequences`, run the `analyse_protein()` function and get the properties.\n",
    "   3. Then the properties returned will be combined into a result variable specific for each protein.\n",
    "   4. Lastly combine all results from all proteins and convert into a `pandas` data frame and return it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a5cc9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyse_multiple_proteins(insulin_fasta):\n",
    "    SeqIO.read(insulin_fasta)\n",
    "    sequences = SeqIO.parse(insulin_fasta)\n",
    "    \n",
    "    for seqeunce in sequences:\n",
    "        analyse_protein(sequence)\n",
    "    \n",
    "def analyse_protein(record):\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cdc8cde3-b238-4a46-9c0d-77be0a279672",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block after function definition on line 3 (792790863.py, line 6)",
     "output_type": "error",
     "traceback": [
      "  \u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[5]\u001b[39m\u001b[32m, line 6\u001b[39m\n\u001b[31m    \u001b[39m\u001b[31mprot_family_analysis = analyze_multiple_proteins('insulin_family.fasta')\u001b[39m\n    ^\n\u001b[31mIndentationError\u001b[39m\u001b[31m:\u001b[39m expected an indented block after function definition on line 3\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def analyze_multiple_proteins(fasta_file):\n",
    "    ## your code here\n",
    "\n",
    "prot_family_analysis = analyze_multiple_proteins('insulin_family.fasta')\n",
    "prot_family_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67e2d303-6394-4f7b-9a40-6ff85a09048b",
   "metadata": {},
   "source": [
    "4. We could use this information to further provide visualizations to compare, study and analyze."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb9ee34-e758-4858-873b-1a092ef76468",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, axes = plt.subplots(2, 2, figsize=(16,10))\n",
    "\n",
    "# Plot 1 (upper left)\n",
    "# Length comparison\n",
    "axes[0,0].bar(prot_family_analysis['Protein'], prot_family_analysis['Length'])\n",
    "axes[0,0].set_title(\"Comparison of Sequence Length\")\n",
    "axes[0,0].set_ylabel('Length (no. amino acids)')\n",
    "axes[0,0].set_xlabel('Protein')\n",
    "axes[0,0].set_ylim(0,max(prot_family_analysis['Length'])+10)\n",
    "axes[0,0].tick_params(axis='x', rotation=45)\n",
    "\n",
    "# Plot 2 (upper right)\n",
    "# Molecular weight\n",
    "axes[0,1].bar(prot_family_analysis['Protein'], prot_family_analysis['Molecular Weight'], color='purple')\n",
    "axes[0,1].set_title(\"Comparison of Molecular Weights\")\n",
    "axes[0,1].set_ylabel('Molecular weights (Da)')\n",
    "axes[0,1].set_xlabel('Protein')\n",
    "axes[0,1].set_ylim(0,max(prot_family_`analysis['Molecular Weight'])+2000)\n",
    "axes[0,1].tick_params(axis='x', rotation=45)\n",
    "\n",
    "# Plot 3 (lower left)\n",
    "# Isoelectric point\n",
    "axes[1,0].bar(prot_family_analysis['Protein'], prot_family_analysis['Isoelectric Point'],color='orange')\n",
    "axes[1,0].set_title(\"Comparison of Isoelectric point\")\n",
    "axes[1,0].set_ylabel('pH')\n",
    "axes[1,0].set_xlabel('Protein')\n",
    "axes[1,0].set_ylim(0,max(prot_family_analysis['Isoelectric Point'])+1)\n",
    "axes[1,0].tick_params(axis='x', rotation=45)\n",
    "\n",
    "# Plot 4 (lower right)\n",
    "# Gravy\n",
    "axes[1,1].bar(prot_family_analysis['Protein'], prot_family_analysis['Gravy'], color='green')\n",
    "axes[1,1].set_title(\"Hydropathy (GRAVY)\")\n",
    "axes[1,1].set_ylabel('Score')\n",
    "axes[1,1].set_xlabel('Protein')\n",
    "axes[1,1].set_ylim(0,max(prot_family_analysis['Gravy'])+0.05)\n",
    "axes[1,1].axhline(y=0, color='r', linestyle='--', linewidth=4)\n",
    "axes[1,1].tick_params(axis='x', rotation=45)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f88eae7-7e4d-4d9c-b454-be34a4f73d5b",
   "metadata": {},
   "source": [
    "What demonstrated above is one of the example how we can extract, compare different protein sequences on the properties of the proteins. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6b483a0-54c3-4863-88a8-3ec32e25e9e1",
   "metadata": {},
   "source": [
    "####\n",
    "---\n",
    "## PART 2: # PART 2: Working with Protein Structures (Intermediate)\n",
    "\n",
    "### 2.1 Understanding PDB Format\n",
    "\n",
    "### Background:\n",
    "The Protein Data Bank (PDB) is a database of 3D structural data of proteins. PDB files contain atomic coordinates in 3D space.\n",
    "\n",
    "### Use Case 3: Analyzing Hemoglobin Structure\n",
    "**Biological Context:** Hemoglobin carries oxygen in blood. Its 3D structure is crucial for function.\n",
    "\n",
    "**Data Source:** RCSB PDB (https://www.rcsb.org/)\n",
    "- Protein: Human Hemoglobin (PDB ID: 1A3N)\n",
    "- Format: PDB file (text format with 3D coordinates)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a49ce9c4-3825-4f4b-abb6-24ad34d6da92",
   "metadata": {},
   "source": [
    "### 2.2 Downloading PDB Files\n",
    "Different with protein sequence, protein structures having more comprehensive information related to the structure of the proteins, location of the amino acids, secondary structure information and etc.\n",
    "\n",
    "#### Method 1: using `requests`\n",
    "The following example shows the way to download the protein structure from Protein Data Bank (RCSB PDB). Basically, the method uses `requests` and uses a standard URL to retrieve the structure file for specific proteins using teh 4-character PDB identifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c75afc-2ea7-4d0f-902c-cb4096818551",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "base_url = \"https://files.rcsb.org/download/\"\n",
    "pdb_id   = \"1A3N\"\n",
    "filename = pdb_id+\".pdb\"\n",
    "req_url  = base_url+filename\n",
    "\n",
    "response = requests.get(req_url)\n",
    "\n",
    "if response.status_code==200:\n",
    "    with open(filename,'w') as file:\n",
    "        file.write(response.text)\n",
    "    print(f'PDB file {filename} downloaded successfully.')\n",
    "else:\n",
    "    print(f'Failed to download. Status code: {response.status_code}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fdef652-7bcf-4725-829b-759523caf190",
   "metadata": {},
   "source": [
    "#### Method 2: using `PDB` module in `Biopython`\n",
    "The following example uses `PDB` module to retrieve the protein structure. However the standard file name are with the extension of `.ent`, but the information are as same as if retrieve using method 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a44ecfd-8481-4c02-9366-aa4082adcff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import PDB\n",
    "\n",
    "pdb_id = \"1A3N\"  # Example PDB ID\n",
    "download_directory = \".\" # <- . means download to the current folder\n",
    "file_format = \"pdb\" \n",
    "\n",
    "pdb_list = PDB.PDBList()\n",
    "\n",
    "pdb_filename = pdb_list.retrieve_pdb_file(pdb_id, pdir=download_directory, file_format=file_format, overwrite=True)\n",
    "print(f\"Downloaded PDB file: {pdb_filename}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60630d51-f990-43bc-8b00-0a4b4e42f8fd",
   "metadata": {},
   "source": [
    "---\n",
    "### 2.3 Parsing PDB Files\n",
    "PDB files are text files with a specific format, but they're not immediately usable for computation. Parsing transforms raw structural data into computable objects.\n",
    "\n",
    "Parsing creates a hierarchical data structure:\n",
    "```\n",
    "Structure\n",
    "  └── Model (experimental structure, usually just one)\n",
    "       └── Chain A, B, C, D... (protein subunits)\n",
    "            └── Residue 1, 2, 3... (amino acids)\n",
    "                 └── Atom N, CA, C, O... (individual atoms)\n",
    "```\n",
    "\n",
    "To parse the PDB files, we will use `PDBParser` function from `PDB` module in `Biopython`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4f11f04-1ef4-4d73-9283-1429027fa7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import PDBParser\n",
    "\n",
    "source = 'pdb1a3n.ent' # the pdb file\n",
    "\n",
    "parser = PDBParser(QUIET=True)\n",
    "structure = parser.get_structure(pdb_id, source)\n",
    "\n",
    "print(\"Structure Information:\")\n",
    "print(\"======================\")\n",
    "print(f\"PDB ID: {structure.id}\")\n",
    "print(f\"Number of models: {len(structure)}\")\n",
    "\n",
    "total_residues = 0\n",
    "total_atoms = 0\n",
    "\n",
    "for ind, model in enumerate(structure):\n",
    "    print(\"\\n----------------------\")\n",
    "    print(f\"Structure for model {ind+1}\")\n",
    "    print(\"----------------------\")\n",
    "    print(f\"Number of chains: {len(model)}\")\n",
    "    print(\"Per-Chain Information\")\n",
    "    for  chain in model:\n",
    "        residues = [residue for residue in chain if residue.id[0]==' '] # checking the residue id if empty means standard amino acids\n",
    "        atoms    = [atom for residue in residues for atom in residue]\n",
    "        residue_count = len(residues)\n",
    "        atom_count    = len(atoms)\n",
    "        print(f'  Chain {chain.id}: {residue_count} residues, {atom_count} atoms')\n",
    "        total_residues += residue_count\n",
    "        total_atoms += atom_count\n",
    "\n",
    "print(f'\\nTotal residues: {total_residues}')\n",
    "print(f'Total atoms: {total_atoms}')\n",
    "print(\"----------------------\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cafb4f0d-9b7b-4d95-9711-30ab2d4afc0c",
   "metadata": {},
   "source": [
    "---\n",
    "#### [Practice]\n",
    "Create a function called `parse_pdb(pdb_file)` that takes the PDB file as input. The function should return the structure parsed from the file, and will display the structure information as above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bdcfb62-cb64-4b4a-8730-456e7205f7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_pdb(pdb_file):\n",
    "    # your code here\n",
    "\n",
    "    return structure\n",
    "\n",
    "# your function should work with this eventually\n",
    "ans = parse_pdb('1A3N.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "411cb1b7-d963-4163-9ad6-ffa128b7cafd",
   "metadata": {},
   "source": [
    "---\n",
    "### 2.4 Calculate Structural Properties\n",
    "\n",
    "#### Center of Mass (COM)\n",
    "The center of mass is like the average position of all the atoms, but weighted by their mass.\n",
    "It tells you where the protein would “balance” if you could somehow place it on a tiny point in 3D space.\n",
    "\n",
    "If all atoms had the same mass, then the center of mass would simply be the average of all (x, y, z) coordinates.\n",
    "But since heavier atoms (like sulfur or oxygen) contribute more, we compute a weighted average instead.\n",
    "\n",
    "The average position of all atoms, weighted by their mass:\n",
    "```\n",
    "COM = Σ(mass_i × position_i) / Σ(mass_i)\n",
    "```\n",
    "\n",
    "Example Pseudocode\n",
    "```\n",
    "1. for each chain in the model\n",
    "2. for each residue in the chain\n",
    "3.    check if the first letter in the residue id is '  ':\n",
    "4.        for each atom in the residue\n",
    "5.            get the mass of that atom\n",
    "6.            get the coordinate of that atom\n",
    "7.            calculate the weighted coordinates\n",
    "8.            calculate the total mass\n",
    "9.        end for\n",
    "10    end if\n",
    "11 end for\n",
    "12 end for\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b7e2d2d-0844-406b-b660-e49fd71c374b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "model  = structure[0] #the structure get from cell above that reads the PDB file\n",
    "\n",
    "total_mass = 0\n",
    "weighted_coords = np.array([0.0, 0.0, 0.0])\n",
    "\n",
    "# --------------\n",
    "# Your code here\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# --------------\n",
    "\n",
    "center = weighted_coords / total_mass\n",
    "print(center)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d205a9-467c-429c-872c-9046bca703da",
   "metadata": {},
   "source": [
    "#### Radius of Gyration (Rg)\n",
    "If the center of mass (COM) tells you where the protein is located in space,\n",
    "the radius of gyration (Rg) tells you how spread out its atoms are around that center. In another words Rg measures how compact or extended the protein structure is.\n",
    "\n",
    "Measure of how spread out the protein is from its center of mass:\n",
    "```\n",
    "Rg = √(Σ(mass_i × distance²_i) / Σ(mass_i))\n",
    "```\n",
    "where distance_i = distance from atom i to COM\n",
    "\n",
    "##### **Physical Interpretation**\n",
    "- **Small Rg** (compact): Protein is tightly packed\n",
    "- **Large Rg** (extended): Protein is spread out\n",
    "- **Intermediate Rg**: Typically folded but with flexible regions\n",
    "\n",
    "Example Pseudocode\n",
    "```\n",
    "1. for each chain in the model\n",
    "2. for each residue in the chain\n",
    "3.    check if the first letter in the residue id is '  ':\n",
    "4.        for each atom in the residue\n",
    "5.            get the mass of that atom\n",
    "6.            get the coordinate of that atom\n",
    "7.            calculate the distance of the atom to from the center\n",
    "8.            calculate the weighted distance\n",
    "9.            calculate the total mass\n",
    "9.        end for\n",
    "10    end if\n",
    "11 end for\n",
    "12 end for\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "961d81ca-9a65-4034-a0c9-79e060ffc57a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = structure[0]\n",
    "\n",
    "total_mass = 0\n",
    "weighted_dist = 0\n",
    "\n",
    "# --------------\n",
    "# Your code here\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# --------------\n",
    "\n",
    "rg = np.sqrt(weighted_dist / total_mass)\n",
    "print(rg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda8901d-3a07-438b-97ac-f027e5ee29b8",
   "metadata": {},
   "source": [
    "##### **alternative**\n",
    "One also can use another library `neurosnap` to get the structure information for a protein (including COM, distance from COM, and more). Though Rg is not directly obtainable but can be calculated based on the distance info obtained from the library. \n",
    "\n",
    "Learn more: https://neurosnap.ai/blog/post/understanding-the-radius-of-gyration-in-protein-design-bioinformatics/66fb0ecc5a941760bf4b5138"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "036789ff-a250-4b18-aa23-0691f2347fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install neurosnap  # uncomment to install it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2934094a-f798-43a7-9a6b-2f07cc03669b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from neurosnap.protein import Protein\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "protID = '1A3N'\n",
    "prot_structure = Protein(protID)\n",
    "\n",
    "\n",
    "print(f\"Structure Information for {protID} using neurosnap\")\n",
    "print(f\"Center of mass: {prot_structure.calculate_center_of_mass()}\")\n",
    "print(f\"Distances from COM: {prot_structure.distances_from_com()}\")\n",
    "\n",
    "R_g = np.sqrt(np.sum(distances_from_com**2) / len(distances_from_com))\n",
    "print(f\"Radius of Gyration of the Protein: {R_g:.2f} Å\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "878aec77-2e2c-4746-9453-390a5160ee33",
   "metadata": {},
   "source": [
    "---\n",
    "### 2.5 Visualizing the protein using `py3Dmol`\n",
    "Using `py3Dmol` we can view the 3D model of a given protein (PDB id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dde7682-3cb6-499d-8f76-806f4aac9e42",
   "metadata": {},
   "outputs": [],
   "source": [
    "import py3Dmol\n",
    "\n",
    "view = py3Dmol.view(query='pdb:1A3N')\n",
    "view.setStyle({'cartoon':{'color':'spectrum'}})\n",
    "view.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca222b61-d755-4654-99c5-1d0173cabb8e",
   "metadata": {},
   "source": [
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
